### ISSUE TITLE
Grails 3 - build fails for domain objects with composite keys

### ISSUE THREAD

**Issue by tcdavid** [longwa/build-test-data#68]
We use the Build Test Data Plugin extensively in our tests for a Grails 2.5.3 app.. When using Grails 3.1.x,  invoking the build() method for domain objects that have composite keys seems to break.  For example, we have a domain class named AccountGroup, which not surprisingly, has references to an Account object and a Group object. It doesn't have its own primary key - it uses a composite of ids for the account and the group objects. When I try to use AccountGroup.build(...) in an integration test, the call ends failing with a NullPointerException in the org.grails.datastore.gorm.GormInstanceApi.ident() method.  The implementation for that method looks like this:

protected Serializable ident(D instance) {
      (Serializable)instance[persistentEntity.getIdentity().name]
  }

It looks like the ident() method needs to protect against nulls returned by the getIdentity() method.

Is there a way for the plugin to handle this correctly?

**Comment by longwa** [comment]
So the NPE is happening in the Grails code? I guess BTD would have to avoid calling this method. I'll have to create a 3.x test scenario (we are still on 2.5.5) and try to workaround it.

**Comment by tcdavid** [comment]
Yes, the NPE happend in the Grails code, in the org.grails.datastore.gorm.GormInstanceApi.ident() method. It looks like they modified the code for Grails 3.2.x, so the issue does not occur for that Grails version. It's still broken when using Grails 3.1.x.  As a workaround for a Grails 3.1.x project, I dropped in a modified version of the GormInstanceApi class with this implementation of the ident() method:

    protected Serializable ident(D instance) {
        def identity = persistentEntity.getIdentity()
        if (identity) {
            return (Serializable) instance[identity.name]
        } else {
            return null
        }
    }

Kind of a kludge, but it seemed to work. Then I found out the Grails 3.2.x version fixed the issue also, so I've moved up to that version of Grails. It would be great if the build plugin still worked with Grails 3.1.
